### PR TITLE
[GIT PULL] Add compile-time and run-time version check API

### LINKED ISSUE #736: version checking macros
It would be really handy if there were version checking macros included with liburing so that code consuming it can do things like:

```c
#if IO_URING_CHECK_VERSION(2,3)
  /* >= 2.3 */
#else
  /* < 2.3 */
#endif
```

I try not to file feature requests of others unless I'm willing to sign up to do the work, so if this is something you want, I'll take the time to do it within reason.

We have a very common pattern for this in GNOME for glib/gtk and associated libraries in the stack.

### PR DESCRIPTION
**Pull request by chergert**
Pretty standard API additions for shared library version checking at both compile-time and run-time.

Adds the following to the ABI for runtime checks:

```
int io_uring_major_version (void);
int io_uring_minor_verison (void);
bool io_uring_check_version(int major, int minor);
```

Adds the following to the API for compile time checks:

```
#define IO_URING_CHECK_VERSION(major,minor) ...
#define IO_URING_VERSION_MAJOR ...
#define IO_URING_VERSION_MINOR ...
```

A new `io_uring_version.h` header is installed next to `compat.h` and generated similarly in `configure`.

Includes some basic tests as well.

The only sort of "wizardry" here is that we use a not-so-common trick to parse `Makefile.common` with `make` to extract the values of `$VERSION_MAJOR` and `$VERSION_MINOR`. (I've been doing this in the GNOME Builder IDE to extract compiler flags for a very, very long time).

Fixes #736

----
## git request-pull output:
```
The following changes since commit 919755a7d0096fda08fb6d65ac54ad8d0fe027cd:

  Add test cases for eventfd and epoll loops (2022-11-30 10:53:09 -0700)

are available in the Git repository at:

  git@github.com:chergert/liburing.git master

for you to fetch changes up to 537107c807e9d307e337cc9495a0efb051b8e350:

  version: add runtime and compile time version checks (2022-11-30 15:52:55 -0800)

----------------------------------------------------------------
Christian Hergert (3):
      Makefile.common: include VERSION_MINOR from specfile
      version: generate io_uring_version.h from Makefile.common
      version: add runtime and compile time version checks

 .gitignore             |  1 +
 Makefile.common        |  1 +
 configure              | 16 ++++++++++++++++
 src/Makefile           |  4 +++-
 src/include/liburing.h | 19 +++++++++++++++++++
 src/version.c          | 21 +++++++++++++++++++++
 test/Makefile          |  1 +
 test/version.c         | 25 +++++++++++++++++++++++++
 8 files changed, 87 insertions(+), 1 deletion(-)
 create mode 100644 src/version.c
 create mode 100644 test/version.c

```
----
## By submitting this pull request, I acknowledge that:
* [x] I have followed the above pull request guidelines.
* [x] I have the rights to submit this work under the same license.
* [x] I agree to a Developer Certificate of Origin (see https://developercertificate.org for more information).

### CODE REVIEW

File: src/include/liburing.h [hunk 1]
```diff
@@ -1348,6 +1348,24 @@ struct io_uring_sqe *io_uring_get_sqe(struct io_uring *ring);
...
+ *
+ * Use IO_URING_CHECK_VERSION() for compile time checks including from
+ * preprocessor directives.
+ *
+ * Use io_uring_check_version() for runtime checks of the version of
+ * liburing that was loaded by the dynamic linker.
+ */
+int io_uring_major_version (void);
+int io_uring_minor_version (void);
+bool io_uring_check_version (int major, int minor);
```
Comment (ammarfaizi2) [inline src/include/liburing.h:1362]: Please remove the space between the function name and arguments list.

File: src/version.c [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+{
+	return IO_URING_VERSION_MINOR;
+}
+
+bool io_uring_check_version (int major, int minor)
+{
+	return major > io_uring_major_version() ||
+		(major == io_uring_major_version() &&
+		 minor >= io_uring_minor_version());
+}
```
Comment (ammarfaizi2) [inline src/version.c:21]: Please remove the space between the function name and arguments list.

**Review by ammarfaizi2 — CHANGES_REQUESTED** [review]
Since you export 3 new functions.

You need add them into the liburing.map as well.

**Comment by axboe** [comment]
It's the echo -e, sh doesn't support that. I wonder how that worked for you?

**Comment by axboe** [comment]
Following incremental makes it work for me:

```
diff --git a/configure b/configure
index 6f444eb96767..82636bb8aa30 100755
--- a/configure
+++ b/configure
@@ -429,9 +429,10 @@ echo "CXX=$cxx" >> $config_host_mak
 print_config "CXX" "$cxx"
 
 # generate io_uring_version.h
+ECHO=$(which echo)
 MAKE_PRINT_VARS="include Makefile.common\nprint-%: ; @echo \$(\$*)\n"
-VERSION_MAJOR=$(echo -e "$MAKE_PRINT_VARS" | make -s -f - print-VERSION_MAJOR)
-VERSION_MINOR=$(echo -e "$MAKE_PRINT_VARS" | make -s -f - print-VERSION_MINOR)
+VERSION_MAJOR=$($ECHO -e "$MAKE_PRINT_VARS" | make -s -f - print-VERSION_MAJOR)
+VERSION_MINOR=$($ECHO -e "$MAKE_PRINT_VARS" | make -s -f - print-VERSION_MINOR)
 io_uring_version_h="src/include/liburing/io_uring_version.h"
 cat > $io_uring_version_h << EOF
 /* SPDX-License-Identifier: MIT */
 ```

**Comment by axboe** [comment]
This is probably better:

```
diff --git a/configure b/configure
index 6f444eb96767..7d8dbf86f322 100755
--- a/configure
+++ b/configure
@@ -430,8 +430,8 @@ print_config "CXX" "$cxx"
 
 # generate io_uring_version.h
 MAKE_PRINT_VARS="include Makefile.common\nprint-%: ; @echo \$(\$*)\n"
-VERSION_MAJOR=$(echo -e "$MAKE_PRINT_VARS" | make -s -f - print-VERSION_MAJOR)
-VERSION_MINOR=$(echo -e "$MAKE_PRINT_VARS" | make -s -f - print-VERSION_MINOR)
+VERSION_MAJOR=$(env echo -e "$MAKE_PRINT_VARS" | make -s -f - print-VERSION_MAJOR)
+VERSION_MINOR=$(env echo -e "$MAKE_PRINT_VARS" | make -s -f - print-VERSION_MINOR)
 io_uring_version_h="src/include/liburing/io_uring_version.h"
 cat > $io_uring_version_h << EOF
 /* SPDX-License-Identifier: MIT */
 ```

**Comment by axboe** [comment]
With that folded in, I think the only missing piece is a man page for the version helpers.

**Comment by chergert** [comment]
Pulled in the `env ` addition for echo and rebased, along with adding man pages for the functions and macros.

Thanks for the review!

**Comment by axboe** [comment]
Thanks! Pulled.

**Comment by YoSTEALTH** [comment]
Thanks @chergert  for added these functions. https://github.com/axboe/liburing/issues/418

Just curious why the functions name starts with `io_uring_*` shouldn't it be `liburing_*` or `uring_*`

**Pull request merged**